clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java

```diff
@@ -16,134 +16,253 @@
+        this.recordSizeMax = createTemplate("record-size-max", METRIC_GROUP_NAME, "The maximum record size", tags);
+        this.recordSizeAvg = createTemplate("record-size-avg", METRIC_GROUP_NAME, "The average record size", tags);
+        this.requestsInFlight = createTemplate("requests-in-flight", METRIC_GROUP_NAME, "The current number of in-flight requests awaiting a response.", tags);
+        this.metadataAge = createTemplate("metadata-age", METRIC_GROUP_NAME, "The age in seconds of the current producer metadata being used.", tags);
+        this.batchSplitRate = createTemplate("batch-split-rate", METRIC_GROUP_NAME, "The average number of batch splits per second", tags);
+        this.batchSplitTotal = createTemplate("batch-split-total", METRIC_GROUP_NAME, "The total number of batch splits", tags);
+
+        this.produceThrottleTimeAvg = createTemplate("produce-throttle-time-avg", METRIC_GROUP_NAME, "The average time in ms a request was throttled by a broker", tags);
+        this.produceThrottleTimeMax = createTemplate("produce-throttle-time-max", METRIC_GROUP_NAME, "The maximum time in ms a request was throttled by a broker", tags);
 
         /* ***** Topic level *****/
-        Set<String> topicTags = new HashSet<String>(tags);
-        topicTags.add("topic");
+        this.topicTags = new HashSet<String>(tags);
+        this.topicTags.add("topic");
+
+        this.topicRecordSendRate = createTemplate("record-send-rate", TOPIC_METRIC_GROUP_NAME, "The average number of records sent per second for a topic.", topicTags);
+        this.topicRecordSendTotal = createTemplate("record-send-total", TOPIC_METRIC_GROUP_NAME, "The total number of records sent for a topic.", topicTags);
+        this.topicByteRate = createTemplate("byte-rate", TOPIC_METRIC_GROUP_NAME, "The average number of bytes sent per second for a topic.", topicTags);
+        this.topicByteTotal = createTemplate("byte-total", TOPIC_METRIC_GROUP_NAME, "The total number of bytes sent for a topic.", topicTags);
+        this.topicCompressionRate = createTemplate("compression-rate", TOPIC_METRIC_GROUP_NAME, "The average compression rate of record batches for a topic.", topicTags);
+        this.topicRecordRetryRate = createTemplate("record-retry-rate", TOPIC_METRIC_GROUP_NAME, "The average per-second number of retried record sends for a topic", topicTags);
+        this.topicRecordRetryTotal = createTemplate("record-retry-total", TOPIC_METRIC_GROUP_NAME, "The total number of retried record sends for a topic", topicTags);
+        this.topicRecordErrorRate = createTemplate("record-error-rate", TOPIC_METRIC_GROUP_NAME, "The average per-second number of record sends that resulted in errors for a topic", topicTags);
+        this.topicRecordErrorTotal = createTemplate("record-error-total", TOPIC_METRIC_GROUP_NAME, "The total number of record sends that resulted in errors for a topic", topicTags);
+
+    }
+
+    public MetricName getBatchSizeAvg() {
```

Comment:
We generally avoid `getters` in Kafka.

Reply:
In this case, the getters aren't standard "getters" which just return the internal variable. These getters are actually instantiating the MetricName object.

For many MetricNames, this isn't *entire* necessary since they are the same each time. But, for MetricNames that have tags (like per-topic MetricNames), it's necessary to instantiate them at get-time, because the tag may vary per metric.

Reply:
@ijuma: Thanks for the feedback. Since you made me realize the getters aren't necessary all of the time, and you also pointed out that almost all fields should be final, I came up with a way to get rid of the getters, which also gets rid of a lot of the boilerplate code. I'll upload that shortly, so you can see how it looks.